README.md
# picosite
## Current Version
The current version is 1.2.3 and it was released on the 31st of March 2020. I seem to have forgotten to update this for the past 3 releases.

## Introduction
A secure CMS / site generator, which only knows how to read files you create. It has no code that can write to the system. It just checks for files. To see it in action check out the [picosite](https://picosite.kloshost.online) website which uses most features, and you can see the source of this site in the 'demosite' branch.

In only roughly 750 lines of code!

There are two branches 'master' is stable and so you can clone right away. There is another branch where I iteratively change stuff and see what happens called 'fiddling' because I just fiddle around with stuff.

There are sometimes brances for big feature introductions.

## Installation

### Mininum

Copy page.php to the remote server and create `main.page` and other page files. You can use picocodes to create your site. Use `REFERENCE.txt` (incomplete) to find a reference of picocodes.

### Standard

Copy everything to your remote server for hosting the website. Then move your getprices.php and secure the site, where you will want to make it so the webserver can't serve up `.page` files on the site. You might also want to delete README.txt and REFERENCE.txt to stop versioning. Read README.txt on your computer or preserve it somehow to learn how to configure picosite.

## Upgrading

### settings.php

You may find with the latest 1.2.0 release that we no longer use config.php - despite a principle of backward compatibility I did not understand PHP GLOBAL() properly. Now settings are inherited from a settings.php which is optional. You can find the settings at the top of page.php and explained in the README.txt file.

## Author Bio

Who is Kaizushi?

I a woman who has escaped from the hell on Earth that is Arabia. I live in a nice secular society now, and I live for Agorism. I run Kaizushi's Little Onion Server which I consider to be free of any jurisdiction. It is hosting, shell accounts, a communication service, and more.

You can find me through the Tor network with this address, and see picosite in action...
http://kaizushigdv5mrnz.onion/ - picosite



CHANGELOG
The 28th of October 2019:
	* I have made the source code release not metion Kaizushi's Little Onion Server.
	* I have made it so picosite only displays old prices if a file for them exists.
The 2nd of November 2019:
	* Created a branch for working on the big feature subpage.
	* I have added a subpage feature but not tested it.
	* Subpage feature required a change to the 404 system, also the 404 page in a
	  file was inconsistent with other ways 404s would be done.
	* Traced each call to getPageFile and put a 404 after for each print method
	  and made informative 404s about what kind of file is missing.
	* I have removed a bunch of debugout calls
	* Now there one can set the currency symbol and this means one can use any
	  language as long as getprice.php is using it.

	* I have implemented the picosite picocall for subpages and wrote the infra-
	  structure for it. It seems to work, and picosite has been changed quite a
	  lot so now this branch is being moved into fiddling.
	  [THIS HAS BEEN MERGED FOR UNIT TESTING INTO FIDDLING]

	* While doing subpages I have changed a lot of methods and reduced if statement
	  by making an empty string that is set if a language is selected.
	* I removed some redundant functions and took out their calls.
	* Made it obvious where one sets the base currency in getprices.php with a global.
The 18th of November 2019:
	* Tested so features are not active without picocodes enabling them.
	* One can set the primary language as a config
	* One can now click on the site logo to go back to the main page and this
	  generates the path, and if an alternate language is passed then it
	  makes a link including it.
The 19th of November 2019:
	* We have added template.php and changed page.php, so that if the template
	  exists, the inbuilt template is disabled. Keep in mind the promise and
	  principle that page.php works solo.
	* The logo image can be clicked on taking one back to the main page for a
	  particular language, if one is selected.
The 23rd of November 2019:
	* It now does not show anything in menus and such, when a translation for
	  a file does not exist.
	* Some changes to the template so it includes a call to print a domains
	  list after languages.
The 28th of November 2019:
	* The domains method was writing HTTPS in its link, now it is just HTTP
	  which should be fine if there is a redirect to HTTPS.
The 17th of December 2019:
	* Now one can use a new picocode '%%##holdpub' to prevent a page from
	  showing in menus, lists, and from displaying its content.
The 5th of January 2020:
	* I have introduced a new feature where one can have settings of where
	  the currency data is kept.
The 31st of March 2020:
	* One can now set a 'pagecode' which can be passed through the URL as
	  GET value 'code' and view pages hidden with the 'holdpub' picocode.
The 6th of April 2020:
	* I have added a feature for random quotes to be put in a file that
	can be displayed. Quotes in the 'quote.txt' file are seperated by
	newlines at the end of each.
	* I have integrated the regular expression which limits file names and
	tested it.


template.php
<html><head><title><?php echo printTitle(); echo " :: "; echo SITENAME; ?></title></head>
<style>
body {
        background:url(rainbow_dash.png) fixed no-repeat bottom right;
        background-color: #222222;
        color: #00AA00;
}
a {
        color: #8888DD;
}
table {
        color: #00DD00;
        background-color: #333333;
        border-radius: 15px;
        opacity: 0.5;
}
</style>
<body>
<center>
<?php printLinksLangs(); ?><br>
<?php printDomains("mirrors: ") ?><br>
<a href="<?php echo getLinkMain(); ?>"><img src="<?php echo SITELOGO; ?>" alt="Site logo image"></img></a>
<p><?php printLinkTop(); ?><br><br>
<?php printBlog(True); ?>
<br><br><br>
<table width=750 cellspacing=10 cellpadding=10><tr><td>
<h1><?php printTitle(); ?></h1>
<?php printPageBody(); ?>
</tr></td></table>
<br><br><br><br><br><br>
<p><?php echo date("Y") . " "; echo SITEAUTHOR; echo ", server time " . date("G:i:s d/m/y") . " UTC"; ?>
<p><i>Powered by Kaizu's <a href="https://github.com/kaizushi/picosite"><?php echo SOFTNAME; ?></a>!</i></p>
</body></html>


hostcheck.php
<?php

// THIS IS NOT FINISHED BUT IS UNIT-TESTED TO BE A GOOD START

// This is a service checker which will write to 'data' in the
// site directory about what services are online. This data
// is for picosite to then display this information.

$_config_hc_srvclist = "servicelist.txt";
$_config_hc_srvcdata = "data/services/";
$_config_hc_redirlim = 5;

if (file_exists("settings.php")) include_once("settings.php");

define("SVCLIST", "$_config_hc_srvclist");
define("SVCDATA", "$_config_hc_srvcdata");

global $redir_count;
$redir_count = 0;

function transEnd($instring, $substring) {
        $length = strlen($substring);
        return $length === 0 || (substr($instring, -$length) === $substring);
}

function checkURL($url) {
	return filter_var($url, FILTER_VALIDATE_URL);
}

function checkServices($services, $base = "") {
	global $redir_count;

	$online = [];
	foreach ($services as $svc) {
		echo "Checking service $svc\n";
		$svc = trim(str_replace("\n", "", $svc));
		$elsewhere = [];
		$curl = curl_init();
		curl_setopt($curl, CURLOPT_HEADER, true);
		curl_setopt($curl, CURLOPT_NOBODY, true);
		curl_setopt($curl, CURLOPT_RETURNTRANSFER,1);
		curl_setopt($curl, CURLOPT_TIMEOUT,10);
		curl_setopt($curl, CURLOPT_URL,$svc);
		$output = curl_exec($curl);
		$httpcode = curl_getinfo($curl, CURLINFO_HTTP_CODE);
		if ($httpcode === 200) {
			echo "Response: 200 OK\n";
			if ($base === "") $base = $svc;
			array_push($online, "$svc|$base");
			curl_close($curl);
			continue;
		}
		if (($httpcode === 301) OR ($httpcode === 302)) {
			$redir_count++;
			if ($redir_count == 3) continue;
			echo "Response: 301 or 302\n";
			$header_size = curl_getinfo($curl, CURLINFO_HEADER_SIZE);
			$headers = substr($output, 0, $header_size);
			$headers = explode("\n", $headers);
			curl_close($curl);
			foreach ($headers as $header) {
				$headparts = explode(" ", $header);
				if ($headparts[0] === "Location:") {
					$url = $headparts[1];
					echo "Checking redirect $url\n";
					$arurl = []; array_push($arurl, $url);
					array_push($elsewhere, checkServices($arurl, $svc));
					array_merge($online, $elsewhere);
				}
			}
		}
		$redir_count = 0;
	}
	
	return $online;
}

function getServiceList() {
	$svclist = file_get_contents(SVCLIST);
	$svclist = explode("\n", $svclist);
	return $svclist;
}

function getServDataLocation() {
	$files = scandir(SVCDATA);

	$fileold = 0;
	$filefin = 0;
	foreach ($files as $file) {
		if ($file === "." || $file === "..") continue;
		$fparts = explode(".", $file);

		$filenum = intval($fparts[0]);
		if ($filenum > $fileold) $filefin = $filenum;
		$fileold = $filenum;
	}

	$filefin++;

	return $filefin;
}

function writeGoodServices($services) {
	$file = strval(SVCDATA . getServDataLocation()) . ".dat";

	$f = fopen($file, 'a');
	foreach ($services as $svc) fwrite($f, "$svc\n");
	fwrite($f, "!!READY!!");
}

function hostcheck() {
	$services = getServiceList();
	$services = checkServices($services);
	writeGoodServices($services);
}

hostcheck();

?>


page.php
<?php

// picosite is almost a CMS, its a light way to have a site.

// DO NOT CHANGE THIS FILE OR YOU WILL BREAK UPDATES.

//default settings

$_config_softname = "picosite 1.3.0";
$_config_debugout = false;
$_config_traceoff = true;
$_config_sitename = "A New Picosite";
$_config_sitelogo = "sitelogo.png";
$_config_suteauth = "Anonymous";
$_config_hidecode = "%%NOCODE%%";
$_config_menuleft = "";
$_config_menumidd = " ";
$_config_menurght = "";
$_config_currsymb = "$";
$_config_currdata = "%%NODATA%%"; 
$_config_hc_srvclist = "servicelist.txt";
$_config_hc_srvcdata = "data/";
$_config_hc_srvcinfo = "serviceinfo.txt";


// You can change any of these settings by creating a settings.php
// and declaring the variables there. If you change this, keep in
// mind future versions will overwrite.

if (file_exists("settings.php")) include_once("settings.php");
if (file_exists("parser.php")) include_once("parser.php");

define("SOFTNAME", $_config_softname);
define("DEBUGOUT", $_config_debugout);
define("TRACEOFF", $_config_traceoff);
define("SITENAME", $_config_sitename);
define("SITELOGO", $_config_sitelogo);
define("SITEAUTHOR", $_config_siteauth);
define("PUBCODE", $_config_hidecode);
define("CURRENCY_SYM", $_config_currsymb);
define("CURRENCY_DAT", $_config_currdata);
define("MENULINK_LFT", $_config_menuleft);
define("MENULINK_MID", $_config_menumidd);
define("MENULINK_RGT", $_config_menurght);
define("SERVLIST_LIST", $_config_hc_srvclist);
define("SERVLIST_DATA", $_config_hc_srvcdata);
define("SERVLIST_INFO", $_config_hc_srvcinfo);

$_GET = array_map('strip_tags', $_GET);

if (isset($argv[1])) $_GET["q"] = $argv[1];

global $HIDEPAGE;
$HIDEPAGE = false;

function debugMethodName(){
	if (TRACEOFF) return;
	$e = new Exception();
	$trace = $e->getTrace();
	//position 0 would be the line that called this function so we ignore it
	$last_call = $trace[1];
	print_r($last_call);
}

function getBool($val) {
	if ($val) return "true";
	else return "false";
}

function printDebug($msg) {
	if (DEBUGOUT) {
		echo "picosite DEBUG: " . $msg . "\n";
		file_put_contents("debug.log", $msg . "\n", FILE_APPEND | LOCK_EX);
		echo "<br>";
	}
}

function printDebugArray($msg, $array) {
	printDebug($msg);
        foreach ($array as $key => $value) printDebug($key . " = " . $value);
}

function transFilesyms($string) {
	//dumb method name
	$newstring = preg_replace("/[^A-Za-z0-9.-]/","",$string);
	return $newstring;
}


function transStart($instring, $substring) {
	$length = strlen($substring);
	return (substr($instring, 0, $length) === $substring);
}

function transEnd($instring, $substring) {
	$length = strlen($substring);
	return $length === 0 || (substr($instring, -$length) === $substring);
}

function transStripNode($nodes) {
	$stripped = [];
	foreach ($nodes as $key => $node) {
		$nodesplit = explode('/', $node);
		$snode = $nodesplit[1];
		$snodesplit = explode('.', $snode);
		$node = $snodesplit[0];
		$stripped[$key] = $node;
	}
	return $stripped;
}

function getLinkMain() {
	// this method makes a 'back to main page' link
	// and preserves the selected language
	$link = "/page.php";

	if (isset($_GET['l'])) $link = $link + "?l=" . $_GET['l'];

	return $link;
}

function getPageTitles($files, $subpage = "[NONE]") {
	$newFiles = [];


	foreach ($files as $file) {
		$paths = explode("/", $file);
			

		if (count($paths) == 2) {
			$file = $paths[1];
			$dir = $paths[0];
		}

		$lang = NULL;
		$inslang = "";
		$fn = NULL;

		$parts = explode('.', $file);
		$title = $parts[0];

		
		if ((transEnd($file, ".blog.page") && $dir === "blogs") ||
		    (transEnd($file, ".guide.page") && $dir === "guides") ||
		    (transEnd($file, ".sub.page") && $dir === $subpage)) {
		    	if (count($parts) == 3) {
				$type = $parts[1];
			} elseif ( count($parts) == 5) {
				$type = $parts[3];
			}
		} elseif (transEnd($file, ".page") && $dir === ".") {
			$type = "page";
	
		}

		if (($subpage === "[NONE]") && $type === "sub") {
			continue;
		}

		if (!is_null($_GET['l'])) {
			$lang = $_GET['l'];
			$inslang = "." . $lang . ".trans";
		}
			
		if ($type === "blog") $fn = "blogs/" . $title . $inslang . ".blog.page";
		if ($type === "guide") $fn = "guides/" . $title . $inslang . ".guide.page";
		if ($type === "sub") $fn = $subpage . "/" . $title . $inslang .  ".sub.page";
		
		if ((($type === "page") && transEnd($file, ".page")))
			$fn = $title . $inslang . ".page";
		
		if (!is_null($fn)) { 
			array_push($newFiles, $fn);
		}
	}

	$files = $newFiles;
	$titledRefs = [];

	foreach ($files as $file) {
		$hastitle = false;
		$ispub = true;

		$filesplit = explode(".", $file);
		$node = $filesplit[0];
		$nodesplit = explode("/", $node);

		if (count($nodesplit) == 2) {
			$node = $nodesplit[1];
		}

		if (!file_exists($file)) {
			$titledRefs[$node] = "Missing File";
			continue;
		}

		$page = file_get_contents($file);
		$lines = explode("\n", $page);
		#find if page requires publish picocode

		foreach ($lines as $line) {
			if (transStart($line, "%%##holdpub")) {
				$ispub = false;
				if ($_GET['code'] === PUBCODE && PUBCODE !== "%%NOCODE%%") $ispub = true;
			}
		}
	

		#find the page title
		foreach ($lines as $line) {
			if (transStart($line, "%%##title=")) {
				$sides = explode("=", $line);
				$title = $sides[1];
				$titledRefs[$node] = $title;
				$hastitle = true;
			}
		}

		printDebug("getPageTitles(): \$ispub = " . getBool($ispub));
		if (!$hastitle) $titledRefs[$node] = "Untitled Page";
		if (!$ispub) $titledRefs[$node] = "%%HIDE%%";
	}

	return $titledRefs;
}

function getPriceBitcoin() {
	$datadir = CURRENCY_DAT;

	if (!transEnd($datadir,"/")) $datadir = $datadir . "/";
	return (float) file_get_contents($datadir . "btc-price");
}

function getPriceMonero() {
	$datadir = CURRENCY_DAT;

        if (!transEnd($datadir,"/")) $datadir = $datadir . "/";
	return (float) file_get_contents($datadir . "xmr-price");
}

function getItemList() {
	$items = [];
	$data = file_get_contents("itemlist.txt");
	$lines = explode("\n", $data);
	foreach ($lines as $line) {
		$split = explode(',', $line);
		$name = $split[0];
		$price = $split[1];
		if ($name !== "") $items[$name] = $price;
	}

	return $items;
}

function getItemListOld() {
	$items = [];

	if (file_exists("itemlist-old.txt") === false) {
		$items['NOFILE'] = true;
		return $items;
	}

	$data = file_get_contents("itemlist-old.txt");
	$lines = explode("\n", $data);
	foreach ($lines as $line) {
		$split = explode(',', $line);
		$name = $split[0];

		$price = $split[1];
		if ($name !== "") $items[$name] = $price;
	}

	return $items;
}

function getPageSubpage($file) {
	$subname = "[NOFILE]";

	if (file_exists($file)) {
		$subname = "[NOSUB]";
		$content = file_get_contents($file);
		$lines = explode("\n", $content);

		foreach ($lines as $line) {
			if (transStart($line, "%%##subpg=")) {
				$exploded = explode("=", $line);
				$subname = $exploded[1];
				break;
			}
		}
	} 
	
	return $subname;
}

function getPageFile($subpagedir = "[NONE]") {
	$fn = "";
	$inslang = "";

	if (!is_null($_GET['q'])) $_GET['q'] = transFilesyms($_GET['q']);
	if (!is_null($_GET['l'])) $_GET['l'] = transFilesyms($_GET['l']);
	if (!is_null($_GET['g'])) $_GET['g'] = transFilesyms($_GET['s']);
	if (!is_null($_GET['sp'])) $_GET['sp'] = transFilesyms($_GET['sp']);


	if (is_null($_GET['q']) || $_GET['q'] === "") $_GET['q'] = 'main';

	if (!is_null($_GET['l'])) $inslang = "." . $_GET['l'] . ".trans";

	if (is_null($_GET['g']) && is_null($_GET['b']) && is_null($_GET['sp'])) 
		$fn = $_GET["q"] . $inslang . ".page";

	if (!is_null($_GET['g']) && is_null($_GET['b']) && is_null($_GET['sp'])) {
		$fn = "guides/" . $_GET['g'] . $inslang . ".guide.page";
	}

	if (is_null($_GET['g']) && !is_null($_GET['b']) && is_null($_GET['sp'])) {
		$fn = "blogs/" . $_GET['b'] . $inslang . ".blog.page";
	}
	
	if (is_null($_GET['g']) && is_null($_GET['b']) && !is_null($_GET['sp'])) {
		if ($subpagedir === "[NONE]") {
			$subname = getPageSubpage($_GET['q'] . ".page");
			if ($subname === "[NOFILE]") {
				$fn = $_GET['q'] . ".page"; //if this looks odd it does a 404	
			} else if ($subname === "[NOSUB]") {
				return $subname;
			} else {
				$subpagedir = $subname;
			}
		}

		$fn = $subpagedir . "/" . $_GET['sp'] . $inslang . ".sub.page";
	}
	
	return $fn;
}

function printServices() {
	$showinfo = true;

	if (!file_exists(SERVLIST_LIST)) return;

	if (!file_exists(SERVLIST_INFO)) $showinfo = false;

	$files = scandir(SERVLIST_DATA);
	if (empty($files)) return;

	$newest = 0;
	$selected = "";

	foreach ($files as $file) {
		$content = file_get_contents($file);
		$ready = false;
		foreach ($content as $data) if ($data === "!!READY!!") $ready = true;
		if ($ready) {
			if ($newest == 0) $newest = filectime($file);
			if (filectime($file) > $newest) {
				$newest = filectime($file);
				$selected = $file;
			}
		}
	}

	$services = file_get_contents($selected);
	$services = explode("\n", $services);
	
	$listed = file_get_contents(SERVLIST_LIST);
	$listed = explode("\n", $listed);

	if ($showinfo) {
		$infos = file_get_contents(SERVLIST_INFO);
		$infos = explode("\n", $infos);
	}

	echo "\n<br>";

	foreach ($listed as $item) {
		echo "$item ";

		$online = false;
		foreach ($services as $svc) 
		{
			$service = "";
			$svc = explode("|", $svc);
			if ($svc[1] === "") $service = $svc[0];
			else $service = $svc[1];

			if ($service === $item) $online = true;

		}
		if ($online) echo " (online)";
		else echo " (offline)";

		if ($showinfo) {
			foreach ($showinfo as $info) {
				$info = explode("|", $info);
				if ($info[0] === $item)	echo " - " . $info[1];
			}
		}

		echo "<br>\n";
	}
}

function printQuotes() {
	if (!file_exists('quotes.txt')) {
		echo SITENAME;
		return;
	}

	$quotes = file_get_contents('quotes.txt');
	$quotes = explode("\n", $quotes);

	$quote = "";
	while ($quote === "") {
		$ar = array_rand($quotes);
		$quote = $quotes[$ar];
	}

	echo $quote;
}

function printDomains($msg) {
	if (file_exists("domainlist.txt")) {
		printDebug("printDomians() domain list exists");
		$domains = file_get_contents("domainlist.txt");
		$domains = explode("\n", $domains);
	} else {
		return;
	}

	echo "$msg";
	foreach ($domains as $domain) {
		if ($_SERVER['HTTP_HOST'] === $domain) continue;
		echo '<a href="http://' . $domain . '/' . '">' . $domain . "</a> ";
	}
	echo "\n";
}

function printCoreOut($text) {
	debugMethodName();

	if (is_string($text)) {
		if (preg_match('/\n/', $text)) {
			$text = explode('\n', $text);
		} else {
			echo "$text";
		}
	}

	if (is_array($text)) {
		foreach ($text as $line) {
			echo "$line\n";
		}
	}
}

function printPrice($itemname, $usdprice, $oldprice) {
	$oldmode = true;

	if ($oldprice === "NOFILE") {
		$oldmode = false;
	}

	$btc = getPriceBitcoin();
	$xmr = getPriceMonero();

	$btcp = number_format($usdprice / $btc, 4);
	$xmrp = number_format($usdprice / $xmr, 3);

	$sym = CURRENCY_SYM;

	if ($oldmode) {
		echo "<tr><td width=\"50%\">$itemname</td><td>$sym<strike>$oldprice</strike></td><td>$sym$usdprice</td><td>BTC: $btcp</td><td>XMR: $xmrp</td></tr>\n";
	} else { 
		echo "<tr><td width=\"50%\">$itemname</td><td>$sym$usdprice</td><td>BTC: $btcp</td><td>XMR: $xmrp</td></tr>\n";
	}
}

function printItemListing() {
	if (CURRENCY_DAT === "%%NODATA%%") {
		return;
	}

	$datadir = CURRENCY_DAT;
        if (!transEnd($datadir,'/')) $datadir = $datadir . "/";
	printDebug("printItemListing: the datadir is $datadir");

	$oldmode = true;
	if (!file_exists("itemlist.txt")) {
		http_response_code(404);
		print "<h2>Price list does not exist.</h2>";
		return;
	}
	$items = getItemList();
	$olditems = getItemListOld();
	$timebtc = filemtime("$datadir/btc-price");
	$timexmr = filectime("$datadir/xmr-price");

	if ($olditems['NOFILE'] === true) {
		$oldmode = false;
		unset($olditems["NOFILE"]);
	}

	$sym = CURRENCY_SYM;

	echo "<p><strong>BTC " . date("G:i:s d/m/Y", $timebtc) . ": $sym" . getPriceBitcoin() . "<br>\n";
	echo "XMR " . date("G:i:s d/m/Y", $timexmr) . ": $sym" . getPriceMonero() . "</strong>\n";
	echo "<table>\n";

	if ($oldmode) {
		echo "<tr><td width=\"50%\"><strong>Name of Item</strong></td><td><strong>Old price</strong></td><td>Price</td><td>Bitcoin cost</td><td>Monero cost</td></tr>\n";
	} else {
		echo "<tr><td width=\"50%\"><strong>Name of Item</strong></td><td>Price</td><td>Bitcoin cost</td><td>Monero cost</td></tr>\n";
	}

	$oldprice=0;
	foreach ($items as $name => $price) {
		$exists = False;
		foreach($olditems as $oname => $oprice) {
			if ($oname === $name) {
				$oldprice = $oprice;
				$exists = True;
			}
		}

		if (!$exists) $oldprice = "XX.XX";
		if (!$oldmode) $oldprice = "NOFILE";
		printPrice($name, $price, $oldprice);
	}
	echo "</table>\n";
}

function printBlog($justone = False, $amount = 0, $start = 0) {
	if ($_GET['b'] == null) {
		if (!file_exists("blogs/")) {
			return;
		}

		if (!is_null($_GET['l'])) $blogdir = glob("blogs/*.trans.blog.page");
		else $blogdir = glob("blogs/*.blog.page");

		$blogs = [];
		$x = 0;

		$inslang = "";
		$getlang = "";
		

		if (!is_null($_GET['l'])) {
			$inslang = "." . $_GET['l'] . ".trans";
			$getlang = "&l=" . $_GET['l'];
		}

		foreach ($blogdir as $blog) {
			if ($blog === ".." || $blog === ".") continue;
			if (is_null($_GET['l']) && transEnd($blog, ".trans.blog.page")) continue;

			if (transEnd($blog, ".blog.page")) {
				$fileparts = explode('.', $blog);
				$name = $fileparts[0];
				$fn = $name . $inslang . ".blog.page";
				$ctime = filemtime($fn);
				$blogs[$x + $ctime] = $fn;
				$x++;
			}
		}

		krsort($blogs);


		$titled = getPageTitles($blogs);
		$blogs = transStripNode($blogs);

		
		echo '<table>';
		$x = 0;
		foreach ($blogs as $time => $blog) {
			$time = $time - $x;
			$x++;
			$thetitle = "Untitled Blog";
			$thenode = "";
			foreach ($titled as $node => $title) {
				if ($blog === $node) {
					$thetitle = $title;
					$thenode = $node;
				}
			}

			if ($thetitle === "%%HIDE%%") {
				continue;
			}

			echo '<tr><td>' . date("G:i:s d/m/Y", $time) . '</td>';
			
			if ($justone) echo "<p>Latest blog post...</p>";

			if (($x - $start) >= 0) {
				echo '<td><a href="/page.php?q=' . $_GET['q'] . '&b='
					. $thenode . $getlang . '">' . $thetitle . "</a></td></tr>\n";
			}

			if (($x - $start) == $amount) break;
			if ($justone) break;
		}
		echo "</table>\n";
	} else {
		if (!file_exists("blogs/")) return;
		if ($justone) return;

		$file = getPageFile();

		if (!file_exists($file)) {
			http_response_code(404);
			echo "<h2>Blog not found in the blogs folder</h2>";
		}

		$lines = explode("\n", $file);

		printCoreOut($lines);
	}
}
 
function printGuide() {
	if ($_GET['g'] == null) {
		echo "<ul>\n";
		if (!file_exists("guides/")) {
			http_reponse_code(404);
			print "<h2>Guides folder does not exist.</h2>";
		}
		$guidesdir = scandir("guides/");
		$guides = [];
		foreach ($guidesdir as $guide) {
			if (transEnd($guide, ".trans.guide.page")) {
				if (!is_null($_GET['l'])) {
					$exploded = explode('.', $guide);
					$guidelang = $exploded[1];
					if ($guidelang === $_GET['l']) {
						array_push($guides, 'guides/' . $guide);
					}
				}
			} else if (transEnd($guide, ".guide.page")) {
				if (is_null($_GET['l'])) {
					array_push($guides, 'guides/' . $guide);
				}
			}
		}

		$titled = getPageTitles($guides);

		foreach ($titled as $node => $title) {
			if ($titled === "%%HIDE%%") continue;
			if (is_null($_GET['l'])) {
				echo '<li><a href="/page.php?q=' . $_GET['q'] . '&g=' . $node . '">' . $title . "</a></li>";
			} else {
				echo '<li><a href="/page.php?q=' . $_GET['q'] . '&g=' . $node . '&l=' . $_GET['l'] . '">' . $title . "</a></li>";
			}
			echo "\n";
		}
		echo "</ul>\n";
			
	} else {
		$file = getPageFile();
		$lines = explode("\n", $file);
		printCoreOut($lines);

		echo '<p><a href="/page.php?q=' . $_GET['q'] . '">Back to Guides</a>';
	}
}

function printSubpage($groupname) {
	$subpages = [];

	if (!file_exists($groupname . '/')) {
		http_response_code(404);
		print "<h2>Subpage called when directory does not exist</h2>";
		return;
	}

	if (is_null($_GET['sp'])) {
		$subpagesdir = scandir($groupname . '/');

		$inslang = "";
		$getlang = "";

		if (!is_null($_GET['l'])) {
			$inslang = "." . $_GET['l'] . ".trans";
			$getlang = "&l=" . $_GET['l'];
		}

		foreach($subpagesdir as $subpage) {
			if ($subpage === ".." || $subpage === ".") continue;

			$exploded = explode('.', $subpage);
			$thissub = $exploded[0];
			$fn = $groupname . '/' . $thissub . $inslang . ".sub.page";

			array_push($subpages, $groupname . '/' . $thissub . $inslang . ".sub.page");
		}

		$titled = getPageTitles($subpages, $groupname);
	
		echo "<ul>";
		foreach($titled as $node => $title) {
			if ($title === "%%HIDE%%") continue;
			if ($title === "Missing File") continue;
			echo '<li><a href="/page.php?q=' . $_GET['q'] . '&sp=' . $node . $getlang .  '">' . $title . "</a></li>\n";
		}
		echo "</ul>";
	} else {
		$file = getPageFile($groupname);

		if (!file_exists($file)) {
			http_response_code(404);
			print "<h2>Subpage called but file does not exist</h2>";
			return;
		}

		$grouptitle = getPageTitle($file);
		printFile($file);
		echo '<p><a href="/page.php?q=' . $_GET['q']  . '">Back to ' . $grouptitle . '</a>';
	}	
}

function printFile($file) {
	global $HIDEPAGE;
	
	if ($HIDEPAGE) {
		echo "The following page is currently hidden from view, and is probably
			a work in progress.";
		http_response_code(404);
		return;
	}

	if (file_exists($file)) {
		debugMethodName();
		$content = file_get_contents($file);
		$lines = explode("\n", $content);
		
		foreach ($lines as $line) {
			if (substr($line, 0, 4) !== "%%##") {
				printCoreOut($line);
				echo "\n";
			} else {
				// eventually this will be a function of
				// similar things but for now it just has
				// icnlude, which includes a file
				if (transStart($line, "%%##incld=")) {
					$exploded = explode("=", $line);
					$incl = $exploded[1];
					printFile($incl);
				}
				if (transStart($line, "%%##subpg=")) {
					$exploded = explode("=", $line);
					$subpage = $exploded[1];
					printSubpage($subpage);
				}
				if (transStart($line, "%%##guide")) {
					printGuide();
				}
				if (transStart($line, "%%##blogs")) {
					printBlog();
				}
				if (transStart($line, "%%##services")) {
					printServices();
				}
				if (transStart($line, "%%##price")) {
					if (transStart($line, "%%##price=all")) {
						printDebug("printFile: we are calling printItemList()");
						printItemListing();
					}
				}
			}
		}
	} else {
		http_response_code(404);
		print("<h2>File not found for this page</h2>");
	}
}

function printPageBody() {
	$pagefile = getPageFile();
		
	printFile($pagefile);
}

function printLinkTop() {
	if (!is_null($_GET['l'])) $files = glob("./*.trans.page");
	else $files = glob("./*.page");

	$links = array();

	$links = getPageTitles($files);

	$ordered = array();

	$file = file_get_contents("menulayout.txt");
	$layout = explode("\n", $file);

	$newlinks = [];

	$rend = true;

	foreach ($layout as $node) {
		if (($links[$node] === "") || ($links[$node] == null)) {
			continue;
		}
		$ordered[$node] = $links[$node];
		unset($links[$node]);
	}

	if (count($layout) == 0) $rend = false;

	$iters = count($ordered);
	$incr = 1;
	if ($rend) echo MENULINK_LFT;
	foreach ($ordered as $node => $title) {
		if (!$rend) break;

		if ($title === "%%HIDE%%") {
			$iters--;
			continue;
		}

		if (is_null($_GET['l'])) {
			echo"<a href=\"/page.php?q=$node\">$title</a>";
		} else {
			$lang = $_GET['l'];
			echo "<a href=\"/page.php?q=$node&l=$lang\">$title</a>";
		}

		if ($incr == $iters) break;
		$incr = $incr + 1;

		echo MENULINK_MID;
	}
	if ($rend) echo MENULINK_RGT . "<br>\n";

	$iters=count($links);
	$incr=1;

	echo MENULINK_LFT;
	foreach ($links as $node => $title) {
		if ($title === "%%HIDE%%") {
			$iters--;
			continue;
		}
		if (($node === "") || ($node == null)) {
			continue;
		}

		if (is_null($_GET['l'])) {
			echo "<a href=\"/page.php?q=$node\">$title</a>";
		} else {
			$lang = $_GET['l'];
			echo "<a href=\"/page.php?q=$node&l=$lang\">$title</a>";
		}

		if ($incr == $iters) break;
		$incr = $incr + 1;

		echo MENULINK_MID;
	}
	echo MENULINK_RGT;
}

function printLinksLangs() {
	if (!file_exists("languages.txt")) return;
	
	$config = file_get_contents("languages.txt");
	$relations = explode("\n", $config);
	$langs = [];
	foreach ($relations as $relation) {
		$splitted = explode(',', $relation);
		$code = $splitted[0];
		$name = $splitted[1];
		$langs[$code] = $name;
	}

	echo '<a href="/page.php?q=' . $_GET['q'] . '">English</a> ';

	$subpglink = "";

	if (!is_null($_GET['sp'])) $subpglink = "&sp=" . $_GET['sp'];

	foreach ($langs as $key => $value) {
		echo '<a href="/page.php?q=' . $_GET['q'] . '&l=' . $key . $subpglink . '">' . $value . '</a> ';
	}

	echo "<br>";
}

function getPageTitle($pagefile) {
	$title = "Unnamed Page";

	if (!file_exists($pagefile)) {
		$title = "Page not found";
	} else {
		$page = file_get_contents($pagefile);
		$lines = explode("\n", $page);

		foreach ($lines as $line) {
			if (transStart($line, "%%##title=")) {
				$sides = explode("=", $line);
				$title = $sides[1];
			}
		}

		$hide = false;

		foreach ($lines as $line) {
			if (transStart($line, "%%##holdpub")) {
				$hide = true;
				if ($_GET['code'] === PUBCODE && PUBCODE !== "%%NOCODE%%") $hide = false;
			}
		}

		if ($hide) $title = "%%HIDE%%";
	}

	return $title;
}

function printTitle() {
	global $HIDEPAGE;
	$pagefile = getPageFile();
	$pagetitle = getPageTitle($pagefile);
	if ($pagetitle === "%%HIDE%%") {
		$HIDEPAGE = true;
		$pagetitle = "404: Hidden Page";
	}
	echo $pagetitle;
}

if (file_exists("template.php")) {
	include("template.php");
	exit();
}

?>

<html><head><title><?php echo printTitle(); echo " :: "; echo SITENAME; ?></title></head>
<style>
body {
	background:url(rainbow_dash.png) fixed no-repeat bottom right;
	background-color: #222222;
	color: #00AA00;
} 
a {
	color: #8888DD;
}
table {
	color: #00DD00;
	background-color: #333333;
	border-radius: 15px;
	opacity: 0.5;
}
</style>
<body>
<center>
<?php printLinksLangs(); ?><br>
<?php printDomains("mirrors: ") ?><br>
<a href="<?php echo getLinkMain(); ?>"><img src="<?php echo SITELOGO; ?>" alt="Site logo image"></img></a>
<p><?php printLinkTop(); ?><br><br>
<?php printBlog(True); ?>
<br><br><br>
<table width=750 cellspacing=10 cellpadding=10><tr><td>
<h1><?php printTitle(); ?></h1>
<?php printPageBody(); ?>
</tr></td></table>
<br><br><br><br><br><br>
<p><?php echo date("Y") . " "; echo SITEAUTHOR; echo ", server time " . date("G:i:s d/m/y") . " UTC"; ?>
<p><i>Powered by Kaizu's <a href="https://github.com/kaizushi/picosite"><?php echo SOFTNAME; ?></a>!</i></p>
</body></html>
